jbi/queue.py

```diff
@@ -0,0 +1,315 @@
+
+    @abstractmethod
+    async def get(self, bug_id: int) -> list[QueueItem]:
+        """Retrieve all of the queue items for a specific bug, sorted in
+        ascending order by the timestamp of the payload event.
+        """
+        pass
+
+    @abstractmethod
+    async def get_all(self) -> dict[int, list[QueueItem]]:
+        """Retrive all items in the queue, grouped by bug
+
+        Returns:
+            dict[int, list[QueueItem]]: Returns a dict of
+            {bug_id: list of events}. Each list of events sorted in ascending
+            order by the timestamp of the payload event.
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def size(self) -> int:
+        pass
+
+
+class MemoryBackend(QueueBackend):
+    def __init__(self):
+        self.existing: dict[int, list[QueueItem]] = defaultdict(list)
+
+    def ping(self):
+        return True
+
+    async def clear(self):
+        logger.debug("Clearing queue")
+        self.existing.clear()
+
+    async def put(self, item: QueueItem):
+        bisect.insort_right(
+            self.existing[item.payload.bug.id],
+            item,
+            key=lambda i: i.payload.event.time or 0,
+        )
+
+    async def get(self, bug_id: int) -> list[QueueItem]:
+        # though we're using a defaultdict, check if the bug is in the queue
+        # this way so that we don't create a key for a bug with no items
+        if bug_id not in self.existing:
+            return []
+        items = self.existing[bug_id]
+        if not items:
+            logger.warn("No items for bug %s, but present in queue", bug_id)
+        return items
+
+    async def get_all(self) -> dict[int, list[QueueItem]]:
+        empty_entries = [not items for items in self.existing.values]
+        if empty_entries:
+            logger.warn(
+                "No items for bugs %s, but present in queue", ",".join(empty_entries)
+            )
+        return self.existing
+
+    async def remove(self, bug_id: int, identifier: str):
+        filtered_items = [
+            i for i in self.existing[bug_id] if i.identifier != identifier
+        ]
+        if not len(filtered_items):
+            del self.existing[bug_id]
+            logger.debug(
+                "Removed %s and entry for bug %s from queue", identifier, bug_id
+            )
+        else:
+            self.existing[bug_id] = filtered_items
+            logger.debug("Removed %s from queue for bug %s", identifier, bug_id)
+
+    @property
+    def size(self) -> int:
+        return sum(len(v) for v in self.existing.values())
+
+
+class FileBackend(QueueBackend):
+    def __init__(self, location):
+        self.location = Path(location)
+        self.location.mkdir(parents=True, exist_ok=True)
+
+    def ping(self):
+        try:
+            with tempfile.TemporaryDirectory(dir=self.location) as temp_dir:
+                with tempfile.TemporaryFile(dir=temp_dir) as f:
+                    f.write(b"")
+            return True
```

Comment:
I was thinking, maybe we could also check `self.get_all()`. Because if a malformed JSON file cannot be parsed, then we want to be alerted

Reply:
Things brings up a good point that currently one "bad apple" can spoil the whole queue. We should probably account for this, especially since one of the reasons we've rejected BZ messages in the past was because they didn't conform to our model.

Maybe we could return a dict of `{int, Generator[QueueItem]}` (or whatever the right syntax for generators is)? So when the retry script hits one of these malformed payloads, we can catch the exception and move on to the next bug. This prevents us from knowing this size of the queue though, I think

Reply:
Good idea for the generator. I hadn't thought that it could delegate the exception handling to the consumer.

The size of queue could be return by the backend just counting files, without parsing them.

I believe we really should know the size of the queue, in order to add warning logs when the size is over a configured threshold (eg. 100?)

The heartbeat endpoint would:
- check that the folder is writable
- iterate the whole queue to make sure all events are parseable
- warn if the size of the queue is over a threshold (100?)
- warn if the age of the oldest event is over a threshold (2 days?)

Then we can also imagine high-levels logs for easier debugging (could be part of retry job?)
- log the time window of events (eg. `Queue contains 14 events starting from 20240301`) 
- log the list of projects concerned by the queued events (eg. `Queue contains 14 events for projects FIDDEDE, FXP`), related #567

Reply:
> The size of queue could be return by the backend just counting files, without parsing them.

Yep, good call. I've accounted for that in my latest commits 👍🏻